src/cluster_context/client.rs

```diff
@@ -0,0 +1,80 @@
+use anyhow::{anyhow, Result};
+use k8s_openapi::api::core::v1::{Namespace, Service};
+use k8s_openapi::api::networking::v1::Ingress;
+use wapc_guest as guest;
+
+pub trait Client {
+    /// Get list of namespaces
+    fn namespaces(&self) -> Result<Vec<u8>>;
+
+    /// Get list of ingresses
+    fn ingresses(&self) -> Result<Vec<u8>>;
+
+    /// Get list of services
+    fn services(&self) -> Result<Vec<u8>>;
+}
+
+pub struct WapcClient {}
+
+impl Client for WapcClient {
+    fn namespaces(&self) -> Result<Vec<u8>> {
```

Comment:
silly question, why do we need to use u8 here, and then do the casting, instead of using k8s_openapi from the beginning? Just to not create complicated mocks when using the TestClient?

Reply:
The idea is to have the client (regardless of its implementation) to return the same object `wapc::host_call` returns. In the beginning I tried to use the same signature, but the dynamic trait over the Error was a mess to handle :(

However, this is why the trait defines methods that return `Vec<u8>` instead of `k8s_openapi` objects.

In terms of UX, the developers do not have to deal with the low level `Vec<u8>` when writing the tests. The `TestClient` has public attributes for namespaces, ingresses and services that use the `k8s_openapi` objects. The conversion to the low level vector of bytes is done behind the scenes.

I hope I understood your question and this answer clarifies it. I'm open to other ways to handle that, if this can lead to a better developer experience.

Reply:
Doh, true, the TestClient implementation uses the k8s_openapi objects.
Yep, that was the general question, thanks for the answer!